kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/GenerateIcebergCompatActionUtils.java

```diff
@@ -0,0 +1,240 @@
+   * features that are blocked via `icebergWriterCompatV1` (for example, rowTracking or
+   * deletionVectors).
+   */
+  private static void validateIcebergWriterCompatV1Enabled(Map<String, String> config) {
+    if (!TableConfig.ICEBERG_WRITER_COMPAT_V1_ENABLED.fromMetadata(config)) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "APIs within GenerateIcebergCompatActionUtils are only supported on tables with"
+                  + " '%s' set to true",
+              TableConfig.ICEBERG_WRITER_COMPAT_V1_ENABLED.getKey()));
+    }
+  }
+
+  /**
+   * Throws an exception if `maxRetries` was not set to 0 in the transaction. We restrict these APIs
+   * to require `maxRetries = 0` since conflict resolution is not supported for operations other
+   * than blind appends.
+   */
+  private static void validateMaxRetriesSetToZero(Row transactionState) {
+    if (getMaxRetries(transactionState) > 0) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Usage of GenerateIcebergCompatActionUtils requires maxRetries=0, "
+                  + "found maxRetries=%s",
+              getMaxRetries(transactionState)));
+    }
+  }
+
+  //////////////////////////////////////////////////
+  // Private methods for creating RemoveFile rows //
+  //////////////////////////////////////////////////
+  // I've added these APIs here since they rely on the assumptions validated within
+  // GenerateIcebergCompatActionUtils such as icebergWriterCompatV1 is enabled --> rowTracking is
+  // disabled. Since these APIs are not valid without these assumptions, holding off on putting them
+  // within RemoveFile.java until we add full support for deletes (which will likely involve
+  // generating RemoveFiles directly from AddFiles anyway)
+
+  private static Row createRemoveFileRowWithExtendedFileMetadata(
+      String path,
+      long deletionTimestamp,
+      boolean dataChange,
+      MapValue partitionValues,
+      long size,
+      Optional<DataFileStatistics> stats,
+      StructType physicalSchema) {
+    Map<Integer, Object> fieldMap = new HashMap<>();
+    fieldMap.put(RemoveFile.FULL_SCHEMA.indexOf("path"), requireNonNull(path));
+    fieldMap.put(RemoveFile.FULL_SCHEMA.indexOf("deletionTimestamp"), deletionTimestamp);
+    fieldMap.put(RemoveFile.FULL_SCHEMA.indexOf("dataChange"), dataChange);
+    fieldMap.put(RemoveFile.FULL_SCHEMA.indexOf("extendedFileMetadata"), true);
+    fieldMap.put(
+        RemoveFile.FULL_SCHEMA.indexOf("partitionValues"), requireNonNull(partitionValues));
+    fieldMap.put(RemoveFile.FULL_SCHEMA.indexOf("size"), size);
+    stats.ifPresent(
+        stat ->
+            fieldMap.put(
+                RemoveFile.FULL_SCHEMA.indexOf("stats"), stat.serializeAsJson(physicalSchema)));
+    return new GenericRow(RemoveFile.FULL_SCHEMA, fieldMap);
+  }
+
+  private static Row convertRemoveDataFileStatus(
```

Comment:
nit: these helper methods should be declared in the order they are used; `convertRemoveDataFileStatus ` should come before `createRemoveFileRowWithExtendedFileMetadata`. Reading our code show read and flow like a newspaper